### PR TITLE
o/devicestate: factory reset mode, no encryption

### PR DESCRIPTION
**Pull request by bboozzoo**
The factory reset handler and support for restoring device serial to the newly created filesystem.

### CODE REVIEW

File: overlord/devicestate/handlers_install.go [hunk 1]
```diff
@@ -618,3 +631,251 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
...
+	}
+	if err != nil {
+		return err
+	}
+	logger.Noticef("found %v serial assertions for %v/%v", len(serials), model.BrandID(), model.Model())
+
+	var serialAs *asserts.Serial
+	for _, serial := range serials {
+		maybeCurrentSerialAs, ok := serial.(*asserts.Serial)
+		if !ok {
```
Comment (pedronis) [inline overlord/devicestate/handlers_install.go:824]: we don't need this kind of check

**Review by pedronis — COMMENTED** [review]
thank you, did a pass,  mostly small comments or order of things considerations

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -1095,6 +1097,54 @@ func (m *DeviceManager) ensureInstalled() error {
...
+	var seeded bool
+	err := m.state.Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !seeded {
+		return nil
+	}
+
+	if m.changeInFlight("factory-reset") {
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go:1125]: yes, good point

File: overlord/devicestate/handlers_install.go [hunk 3]
```diff
@@ -618,3 +631,251 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
...
+	}
+
+	// TODO: see notes in install handler
+
+	// on some specific devices, we need to create these directories in
+	// _writable_defaults in order to allow the install-device hook to install
+	// some files there, this eventually will go away when we introduce a proper
+	// mechanism not using system-files to install files onto the root
+	// filesystem from the install-device hook
+	if err := fixupWritableDefaultDirs(boot.InstallHostWritableDir); err != nil {
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/handlers_install.go:775]: similar tweak would be needed in install mode then too, i can do that in a follow up so as to not mix the factory reset and install changes in one PR

File: overlord/devicestate/handlers_install.go (hunk 1 above)
Comment (bboozzoo) [reply overlord/devicestate/handlers_install.go:824]: dropped

File: overlord/devicestate/devicemgr.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go]: applied some of the cleanups for install mode in https://github.com/snapcore/snapd/pull/11563

**Review by pedronis — APPROVED** [review]
thanks

File: overlord/devicestate/devicestate_install_mode_test.go [hunk 4]
```diff
@@ -1636,3 +1639,536 @@ func (s *deviceMgrInstallModeSuite) TestInstallModeWritesTimesyncdClockErr(c *C)
...
+echo "mock output of: $(basename "$0") $*"
+`)
+	defer mockedSnapCmd.Restore()
+
+	// pretend snap-bootstrap mounted ubuntu-save
+	err := os.MkdirAll(boot.InitramfsUbuntuSaveDir, 0755)
+	c.Assert(err, IsNil)
+	// and it has some content
+
+	serial := makeDeviceSerialAssertionInDir(c, boot.InstallHostDeviceSaveDir, s.storeSigning, s.brands,
```
Comment (mardy) [inline overlord/devicestate/devicestate_install_mode_test.go:1803]: nitpick: the comment was probably meant to be on the next line

File: overlord/devicestate/devicestate_install_mode_test.go [hunk 5]
```diff
@@ -1636,3 +1639,536 @@ func (s *deviceMgrInstallModeSuite) TestInstallModeWritesTimesyncdClockErr(c *C)
...
+	c.Assert(err, IsNil)
+	c.Assert(ass, HasLen, 1)
+	asSerial, _ := ass[0].(*asserts.Serial)
+	c.Assert(asSerial, NotNil)
+	c.Assert(asSerial, DeepEquals, serial)
+
+	kp, err := asserts.OpenFSKeypairManager(assertsInResetSystem)
+	c.Assert(err, IsNil)
+	_, err = kp.Get(serial.DeviceKey().ID())
+	// the will not have been found, as this is a device with ubuntu-save
```
Comment (mardy) [inline overlord/devicestate/devicestate_install_mode_test.go:1839]: the "key", I guess

File: overlord/devicestate/export_test.go [hunk 6]
```diff
@@ -328,6 +328,14 @@ func MockInstallRun(f func(model gadget.Model, gadgetRoot, kernelRoot, device st
 	}
 }
 
+func MockInstallFactoryReset(f func(model gadget.Model, gadgetRoot, kernelRoot, device string, options install.Options, observer gadget.ContentObserver, perfTimings timings.Measurer) (*install.InstalledSystemSideData, error)) (restore func()) {
+	old := installFactoryReset
```
Comment (mardy) [inline overlord/devicestate/export_test.go:332]: you could use testutil.Backup()

File: overlord/devicestate/devicestate_install_mode_test.go [hunk 7]
```diff
@@ -1636,3 +1639,536 @@ func (s *deviceMgrInstallModeSuite) TestInstallModeWritesTimesyncdClockErr(c *C)
...
+		}
+	}
+	return nil
+}
+
+func (s *deviceMgrInstallModeSuite) TestFactoryResetExpectedTasks(c *C) {
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	restore = devicestate.MockInstallFactoryReset(func(mod gadget.Model, gadgetRoot, kernelRoot, device string, options install.Options, _ gadget.ContentObserver, _ timings.Measurer) (*install.InstalledSystemSideData, error) {
```
Comment (mardy) [inline overlord/devicestate/devicestate_install_mode_test.go:2058]: nitpick: some variables have a name, others are a `_`.

File: overlord/devicestate/handlers_install.go [hunk 8]
```diff
@@ -618,3 +631,255 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
...
+	bopts := install.Options{
+		Mount: true,
+	}
+	encryptionType, err := m.checkEncryption(st, deviceCtx)
+	if err != nil {
+		return err
+	}
+	bopts.EncryptionType = encryptionType
+	useEncryption := (encryptionType != secboot.EncryptionTypeNone)
+	hasMaker := osutil.FileExists(filepath.Join(boot.InstallHostFDESaveDir, "marker"))
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:677]: "hasMarker"?

File: overlord/devicestate/handlers_install.go [hunk 9]
```diff
@@ -618,3 +631,255 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
...
+	// TODO verify that the same encryption mechanism is used
+	if hasMaker != useEncryption {
+		prevStatus := "encrypted"
+		if !hasMaker {
+			prevStatus = "unencrypted"
+		}
+		// the original system was (un)encrypted
+		return fmt.Errorf("cannot perform factory reset using different encryption, the original system was %v", prevStatus)
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:685]: I think that the error message is more clear than the comment, which I'd rather remove.

File: overlord/devicestate/handlers_install.go [hunk 10]
```diff
@@ -618,3 +631,255 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
...
+	// TODO: splice into a common install/reset helper?
+
+	// keep track of the model we installed
+	err = os.MkdirAll(filepath.Join(boot.InitramfsUbuntuBootDir, "device"), 0755)
+	if err != nil {
+		return fmt.Errorf("cannot store the model: %v", err)
+	}
+	err = writeModel(model, filepath.Join(boot.InitramfsUbuntuBootDir, "device/model"))
+	if err != nil {
+		return fmt.Errorf("cannot store the model: %v", err)
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:725]: do we need to remove the `device` directory we just created? If not, it's still better writing a comment about it.

File: overlord/devicestate/handlers_install.go [hunk 11]
```diff
@@ -618,3 +631,255 @@ func (m *DeviceManager) doRestartSystemToRunMode(t *state.Task, _ *tomb.Tomb) er
...
+		// no serial assertion that matches the model, brand and is
+		// signed with a device key that is present in the filesystem
+		logger.Debugf("no valid serial assertions")
+		return nil
+	}
+
+	logger.Debugf("found a serial assertion for %v/%v, with serial %v",
+		model.BrandID(), model.Model(), serialAs.Serial())
+
+	toDB, err := sysdb.OpenAt(filepath.Join(boot.InstallHostWritableDir, "var/lib/snapd/assertions"))
```
Comment (mardy) [inline overlord/devicestate/handlers_install.go:863]: Do we need to close `toDB`?

**Review by mardy — APPROVED** [review]
LGTM, with a few irrelevant nitpicks (which you can ignore, if you don't need to touch the code again) and questions :-)

File: overlord/devicestate/handlers_install.go (hunk 11 above)
Comment (bboozzoo) [reply overlord/devicestate/handlers_install.go:863]: no, there's no close operation

File: overlord/devicestate/handlers_install.go (hunk 10 above)
Comment (bboozzoo) [reply overlord/devicestate/handlers_install.go:725]: no not really, if this fails the whole factory reset fails and only way out is doing a reinstall which wipes this partition

**Review by mardy — APPROVED** [review]
Thanks!

**Pull request merged**